### PR TITLE
File drop control

### PR DESCRIPTION
**Pull request by Blackbaud-TrevorBurch**

### CODE REVIEW

File: src/app/public/modules/file-attachment/file-drop-control.component.html [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  (filesChanged)="filesUpdated($event)"
+  (linkChanged)="linkUpdated($event)"
+>
+  <ng-content></ng-content>
+</sky-file-drop>
+
+<ng-content select="sky-alert"></ng-content>
+
+<div *ngFor="let item of allItems">
+  <sky-file-item
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/file-attachment/file-drop-control.component.html:19]: Now that we're moving the `sky-file-item`s to this template, how would a consumer load the new control with values already added? I tried to add a `SkyFileDropValue` to the `new FormControl()` statement, but it didn't add any `sky-file-item`s to the template. Do we support programmatically adding files/links?

File: src/app/public/modules/file-attachment/file-drop-control.component.ts [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+  constructor(
+    private changeDetector: ChangeDetectorRef,
+    @Self() @Optional() private ngControl: NgControl
+  ) {
+    if (this.ngControl) {
+      this.ngControl.valueAccessor = this;
+    }
+  }
+
+  public deleteItem(event: SkyFileLink | SkyFileItem) {
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/file-attachment/file-drop-control.component.ts:110]: Some of these methods need a return type.

File: src/app/public/modules/file-attachment/file-drop-control.component.spec.ts [hunk 3]
```diff
@@ -0,0 +1,2957 @@
...
+    });
+  });
+
+  /**
+   * NOTE: Most tests are the same as the file drop on 4/1/2020.
+   * This is to ensure the wrapper does not break existing functionality.
+   * Extra testing has been added to ensure that form controls work correctly
+   */
+
+  describe('reactive forms', () => {
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/file-attachment/file-drop-control.component.spec.ts:70]: I'm not seeing any tests in here for checking the form control's state (dirty, valid, etc...). Should we also test for properly handling `required` and `disabled` states?

File: src/app/public/modules/file-attachment/file-drop-control.component.spec.ts [hunk 4]
```diff
@@ -0,0 +1,2957 @@
...
+      let filesChangedActual: SkyFileDropChange;
+
+      fileDropControl.filesUploaded.subscribe(
+        (filesChanged: SkyFileDropChange) => filesChangedActual = filesChanged);
+
+      fileDropControl.acceptedTypes = 'image/png,image/tiff';
+
+      fixture.detectChanges();
+
+      let files = [
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/file-attachment/file-drop-control.component.spec.ts:516]: It looks like some of these `let`s can be `const`s.

File: src/app/public/modules/file-attachment/fixtures/file-drop-control-no-form.component.fixture.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  SkyFileDropControlComponent
+} from '../file-drop-control.component';
+
+@Component({
+  selector: 'file-drop-control-no-form-test',
+  templateUrl: './file-drop-control-no-form.component.fixture.html'
+})
+export class FileDropControlNoFormTestComponent {
+
+  public required: boolean = false;
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/file-attachment/fixtures/file-drop-control-no-form.component.fixture.ts:16]: It looks like `required` and `showLabel` aren't being used?

File: src/app/public/modules/file-attachment/fixtures/file-drop-control-template.component.fixture.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+@Component({
+  selector: 'file-drop-control-template-test',
+  templateUrl: './file-drop-control-template.component.fixture.html'
+})
+export class FileDropControlTemplateTestComponent {
+
+  public attachment: SkyFileDropValue;
+
+  public required: boolean = false;
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/file-attachment/fixtures/file-drop-control-template.component.fixture.ts:22]: It looks like `required` and `showLabel` aren't being used?

File: src/app/public/modules/file-attachment/fixtures/file-drop-control.compoennt.fixture.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  SkyFileDropControlComponent
+} from '../file-drop-control.component';
+
+@Component({
+  selector: 'file-drop-control-test',
+  templateUrl: './file-drop-control.component.fixture.html'
+})
+export class FileDropControlTestComponent implements OnInit {
+
+  public attachment: FormControl;
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/file-attachment/fixtures/file-drop-control.compoennt.fixture.ts:23]: It looks like `required` and `showLabel` aren't being used?

File: src/app/public/modules/file-attachment/fixtures/file-drop-control.compoennt.fixture.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+import {
+  SkyFileDropControlComponent
+} from '../file-drop-control.component';
+
+@Component({
+  selector: 'file-drop-control-test',
+  templateUrl: './file-drop-control.component.fixture.html'
+})
+export class FileDropControlTestComponent implements OnInit {
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/file-attachment/fixtures/file-drop-control.compoennt.fixture.ts:21]: Typo in the file name: `compoennt`

File: src/app/public/modules/file-attachment/fixtures/file-drop-test-helper.ts [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+
+import {
+  ComponentFixture
+} from '@angular/core/testing';
+
+import {
+  By
+} from '@angular/platform-browser';
+
+export class FileDropTestHelper {
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/file-attachment/fixtures/file-drop-test-helper.ts:13]: Nice helper!

File: src/app/public/modules/file-attachment/file-attachment.component.ts [hunk 10]
```diff
@@ -60,7 +61,7 @@ let uniqueId = 0;
   styleUrls: ['./file-attachment.component.scss'],
   changeDetection: ChangeDetectionStrategy.OnPush
 })
-export class SkyFileAttachmentComponent implements AfterViewInit, AfterContentInit, OnDestroy {
+export class SkyFileAttachmentComponent implements AfterViewInit, AfterContentInit, ControlValueAccessor, OnDestroy {
```
[the remark this answers is not in the document]
Comment (Blackbaud-TrevorBurch) [reply src/app/public/modules/file-attachment/file-attachment.component.ts:64]: It should have had the implements value previously but didn't. This was really a small code cleanup. The functions had been implemented in previous work.

File: src/app/public/modules/file-attachment/file-drop.component.ts [hunk 11]
```diff
@@ -29,6 +29,13 @@ import {
   styleUrls: ['./file-drop.component.scss']
 })
 export class SkyFileDropComponent {
+
+  /** @internal
+   * Used by the file drop control component to handle when the drop itself is touched
+   */
+  @Output()
+  public controlsTouched = new EventEmitter<void>();
```
[the remark this answers is not in the document]
Comment (Blackbaud-TrevorBurch) [reply src/app/public/modules/file-attachment/file-drop.component.ts:37]: The issue is that another component needs to use this and it isn't internal like that lookup one was (note that it is a private variable). I can make it a public subject and not mark it as an output but that will require us to then use a `ViewChild` in the new control where we don't have to do that now and can just utilize the html binding

File: src/app/public/modules/file-attachment/types/file-drop-value.ts [hunk 12]
```diff
@@ -0,0 +1,14 @@
+import {
+  SkyFileItem
+} from '../file-item';
+
+import {
+  SkyFileLink
+} from '../file-link';
+
+export interface SkyFileDropValue {
```
[the remark this answers is not in the document]
Comment (Blackbaud-TrevorBurch) [reply src/app/public/modules/file-attachment/types/file-drop-value.ts:9]: Ya. I'll add those with jsdocs throughout

File: src/app/public/modules/file-attachment/file-drop.component.spec.ts [hunk 13]
```diff
@@ -5,34 +5,35 @@ import {
...
+} from './file-drop.component';
 
 import {
   SkyFileLink
 } from './file-link';
 
+import {
+  SkyFileDropChange
+} from './types/file-drop-change';
+import { FileDropTestHelper } from './fixtures/file-drop-test-helper';
```
[the remark this answers is not in the document]
Comment (Blackbaud-TrevorBurch) [reply src/app/public/modules/file-attachment/file-drop.component.spec.ts:35]: UGH! I thought I managed to check all the files for those..... fixed.

File: src/app/public/modules/file-attachment/file-drop.component.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (Blackbaud-AlexKingman) [reply src/app/public/modules/file-attachment/file-drop.component.ts:37]: Yeah, that's what I was suggesting. I think the `ViewChild` route, while a bit more tedious, doesn't expose an additional `@Output` parameter that a consumer might mistake for part of the API. Also makes documentation more obvious.

File: src/app/public/modules/file-attachment/file-drop-control.component.html [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+  (controlsTouched)="handleControlTouch()"
+  (filesChanged)="filesUpdated($event)"
+  (linkChanged)="linkUpdated($event)"
+>
+  <ng-content></ng-content>
+</sky-file-drop>
+
+<ng-content select="sky-alert"></ng-content>
+
+<div *ngFor="let item of allItems">
```
[the remark this answers is not in the document]
Comment (Blackbaud-TrevorBurch) [reply src/app/public/modules/file-attachment/file-drop-control.component.html:18]: Good call. I honestly pulled the basis for the control off the demo for our docs on the existing component. I didn't notice the extra div. I've removed it now

File: src/app/public/modules/file-attachment/file-drop-control.component.ts [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+
+import {
+  SkyFileDropValue
+} from './types/file-drop-value';
+
+@Component({
+  selector: 'sky-file-drop-control',
+  templateUrl: './file-drop-control.component.html'
+})
+export class SkyFileDropControlComponent implements ControlValueAccessor {
```
[the remark this answers is not in the document]
Comment (Blackbaud-TrevorBurch) [reply src/app/public/modules/file-attachment/file-drop-control.component.ts:36]: Yup. I'll add those. I was actually thinking about that this morning.

File: src/app/public/modules/file-attachment/file-drop-control.component.ts [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+
+  @Input()
+  public acceptedTypes: string;
```
[the remark this answers is not in the document]
Comment (Blackbaud-TrevorBurch) [reply src/app/public/modules/file-attachment/file-drop-control.component.ts:39]: Yup. Will do

File: src/app/public/modules/file-attachment/file-drop-control.component.ts [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+    this.filesUploaded.emit(event);
+    this.value = {
+      files: (this.value && this.value.files) ? this.value.files.concat(event.files) : event.files,
+      links: this.value ? this.value.links : undefined
+    };
+    this.allItems = this.allItems.concat(event.files);
+  }
+
+  public handleControlTouch() {
+    this._onTouched();
```
[the remark this answers is not in the document]
Comment (Blackbaud-TrevorBurch) [reply src/app/public/modules/file-attachment/file-drop-control.component.ts:153]: Yup. I followed what the single file attachment does. I actually looked into this when I was implementing and it was difficult to tell when the file picker was closed too. I think we should mimic what we do in the single file attachment. Thoughts?

File: src/app/public/modules/file-attachment/file-drop-control.component.html (hunk 1 above)
Comment (Blackbaud-TrevorBurch) [reply src/app/public/modules/file-attachment/file-drop-control.component.html:19]: I'll look at this.

File: src/app/public/modules/file-attachment/file-drop-control.component.ts (hunk 2 above)
Comment (Blackbaud-TrevorBurch) [reply src/app/public/modules/file-attachment/file-drop-control.component.ts:110]: Fixed

File: src/app/public/modules/file-attachment/file-drop-control.component.spec.ts (hunk 3 above)
Comment (Blackbaud-TrevorBurch) [reply src/app/public/modules/file-attachment/file-drop-control.component.spec.ts:70]: 1. I'll add some for checking the state

2. We don't yet fully support those. We have an issue for that but need design before we can fully implement it.

File: src/app/public/modules/file-attachment/file-drop-control.component.spec.ts (hunk 4 above)
Comment (Blackbaud-TrevorBurch) [reply src/app/public/modules/file-attachment/file-drop-control.component.spec.ts:516]: Uhhh. Probably lol. I'll look through them. Most tests were taken straight from the existing file drop tests.

File: src/app/public/modules/file-attachment/fixtures/file-drop-test-helper.ts (hunk 9 above)
Comment (Blackbaud-TrevorBurch) [reply src/app/public/modules/file-attachment/fixtures/file-drop-test-helper.ts:13]: Thanks!

File: src/app/public/modules/file-attachment/fixtures/file-drop-control.compoennt.fixture.ts (hunk 8 above)
Comment (Blackbaud-TrevorBurch) [reply src/app/public/modules/file-attachment/fixtures/file-drop-control.compoennt.fixture.ts:21]: Nice catch. Fixed.

File: src/app/public/modules/file-attachment/fixtures/file-drop-control-template.component.fixture.ts (hunk 6 above)
Comment (Blackbaud-TrevorBurch) [reply src/app/public/modules/file-attachment/fixtures/file-drop-control-template.component.fixture.ts:22]: Removed

File: src/app/public/modules/file-attachment/fixtures/file-drop-control.compoennt.fixture.ts (hunk 7 above)
Comment (Blackbaud-TrevorBurch) [reply src/app/public/modules/file-attachment/fixtures/file-drop-control.compoennt.fixture.ts:23]: Removed

File: src/app/public/modules/file-attachment/fixtures/file-drop-control-no-form.component.fixture.ts (hunk 5 above)
Comment (Blackbaud-TrevorBurch) [reply src/app/public/modules/file-attachment/fixtures/file-drop-control-no-form.component.fixture.ts:16]: Removed

File: src/app/public/modules/file-attachment/file-drop-control.component.spec.ts (hunk 4 above)
Comment (Blackbaud-TrevorBurch) [reply src/app/public/modules/file-attachment/file-drop-control.component.spec.ts:516]: *wipes sweat* - Done

File: src/app/public/modules/file-attachment/file-drop-control.component.ts [hunk 18]
```diff
@@ -0,0 +1,180 @@
...
+  public filesUploaded = new EventEmitter<SkyFileDropChange>();
+
+  @Output()
+  public fileDeleted = new EventEmitter<SkyFileItem>();
+
+  @Output()
+  public linkAdded = new EventEmitter<SkyFileLink>();
+
+  @Output()
+  public linkDeleted = new EventEmitter<SkyFileLink>();
```
[the remark this answers is not in the document]
Comment (Blackbaud-TrevorBurch) [reply src/app/public/modules/file-attachment/file-drop-control.component.ts:69]: Done

File: src/app/public/modules/file-attachment/file-drop-control.component.ts [hunk 19]
```diff
@@ -0,0 +1,180 @@
...
+  public multiple: boolean = true;
+
+  @Input()
+  public noClick: boolean = false;
+
+  @Input()
+  public validateFn: Function;
+
+  @Output()
+  public filesUploaded = new EventEmitter<SkyFileDropChange>();
```
[the remark this answers is not in the document]
Comment (Blackbaud-TrevorBurch) [reply src/app/public/modules/file-attachment/file-drop-control.component.ts:60]: Done

**Pull request closed without merging**